File: packages/dart/lib/src/tasklog.dart
/// 任务流水与「走过」的算法。
///
/// 一条流水＝时间、步骤名、一句话、过没过。步骤名带后缀的（`·审` 审查、`·判` 机器判据）
/// **给这一步的结论投票**；不带后缀的（重新执行一次）**把结论从头算**。
String textOf(Object? value, String key) {
  if (value is! Map) return '';
  final item = value[key];
  return item is String ? item.trim() : '';
}

/// 哪些步骤走过了。[steps] 是工作流上的步骤名，按定义顺序。
List<String> done(List<String> steps, List<Map> events) {
  final verdict = <String, bool>{};
  final order = <String>[];
  for (final event in events) {
    final ok = event['ok'] == true;
    final raw = '${event['step'] ?? ''}';
    final at = raw.indexOf('·');
    final step = at < 0 ? raw : raw.substring(0, at);
    final hasExtra = at >= 0;
    if (!steps.contains(step)) continue;
    if (verdict.containsKey(step)) {
      verdict[step] = hasExtra ? (verdict[step]! && ok) : ok;
    } else {
      verdict[step] = ok;
      order.add(step);
    }
  }
  return order.where((step) => verdict[step] == true).toList();
}

/// 第一个没走到的步骤。
String? nextStep(List<String> steps, List<Map> events) {
  final finished = done(steps, events);
  for (final name in steps) {
    if (!finished.contains(name)) return name;
  }
  return null;
}

/// 状态行：下一步是谁，或者都走过了。
String stateLine(List<String> steps, List<Map> events, String workflowName) {
  if (steps.isEmpty) {
    return '这条工作流没有步骤——在 workflows/$workflowName.yaml 的 steps 里写步骤';
  }
  final next = nextStep(steps, events);
  return next == null ? '${steps.length} 个步骤都走过了' : '下一步：$next';
}

File: packages/rust/src/tasklog.rs
//! 任务流水与「走过」的算法。
//!
//! 一条流水＝时间、步骤名、一句话、过没过。步骤名带后缀的（`·审` 审查、`·判` 机器判据）
//! **给这一步的结论投票**；不带后缀的（重新执行一次）**把结论从头算**。

use serde_json::Value as Json;

pub fn text_of(value: &Json, key: &str) -> String {
    value
        .get(key)
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .trim()
        .to_string()
}

/// 哪些步骤走过了。`steps` 是工作流上的步骤名，按定义顺序。
pub fn done(steps: &[String], events: &[Json]) -> Vec<String> {
    let mut verdict: Vec<(String, bool)> = Vec::new();
    for event in events {
        let ok = event.get("ok").and_then(|v| v.as_bool()).unwrap_or(false);
        let raw = text_of(event, "step");
        let (step, extra) = match raw.split_once('·') {
            Some((step, rest)) => (step.to_string(), Some(rest.to_string())),
            None => (raw.clone(), None),
        };
        if !steps.contains(&step) {
            continue;
        }
        match verdict.iter_mut().find(|(name, _)| *name == step) {
            Some((_, last)) => {
                if extra.is_some() {
                    *last = *last && ok;
                } else {
                    *last = ok;
                }
            }
            None => verdict.push((step, ok)),
        }
    }
    verdict
        .into_iter()
        .filter(|(_, ok)| *ok)
        .map(|(name, _)| name)
        .collect()
}

/// 第一个没走到的步骤。
pub fn next_step(steps: &[String], events: &[Json]) -> Option<String> {
    let finished = done(steps, events);
    steps.iter().find(|name| !finished.contains(name)).cloned()
}

/// 状态行：下一步是谁，或者都走过了。
pub fn state_line(steps: &[String], events: &[Json], workflow_name: &str) -> String {
    if steps.is_empty() {
        return format!("这条工作流没有步骤——在 workflows/{workflow_name}.yaml 的 steps 里写步骤");
    }
    match next_step(steps, events) {
        Some(step) => format!("下一步：{step}"),
        None => format!("{} 个步骤都走过了", steps.len()),
    }
}
